### PR TITLE
[Reland] Implement cusparse Descriptor class and clean up cusparse code 

### PR DESCRIPTION
**Pull request by xwang233**
Reland of #37389. Fix for the cuda 10.1 CI failure.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by pbelevich** [comment]
Can not compile pytorch on devfair with the following env:
```
(pytorch-py3.7) pbelevich@devfair0317:/scratch/pbelevich/pytorch$ module list
Currently Loaded Modulefiles:
 1) cuda/10.1   2) cudnn/v7.6.5.32-cuda.10.1
(pytorch-py3.7) pbelevich@devfair0317:/scratch/pbelevich/pytorch$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Fri_Feb__8_19:08:17_PST_2019
Cuda compilation tools, release 10.1, V10.1.105
```

```
-- Found CUDA: /public/apps/cuda/10.1 (found version "10.1")
-- Caffe2: CUDA detected: 10.1
-- Caffe2: CUDA nvcc is: /private/home/pbelevich/ccache/cuda/nvcc
-- Caffe2: CUDA toolkit directory: /public/apps/cuda/10.1
-- Caffe2: Header version is: 10.1
-- Found CUDNN: /public/apps/cudnn/v7.6.5.32/cuda/lib64/libcudnn.so
-- Found cuDNN: v7.6.5  (include: /public/apps/cudnn/v7.6.5.32/cuda/include, library: /public/apps/cudnn/v7.6.5.32/cuda/lib64/libcudnn.so)
-- Autodetected CUDA architecture(s):  6.0 6.0
-- Added CUDA NVCC flags for: -gencode;arch=compute_60,code=sm_60
-- Autodetected CUDA architecture(s):  6.0 6.0
```

```
--   USE_CUDA              : ON
--     CUDA static link    : OFF
--     USE_CUDNN           : ON
--     CUDA version        : 10.1
--     cuDNN version       : 7.6.5
--     CUDA root directory : /public/apps/cuda/10.1
--     CUDA library        : /public/apps/cuda/10.1/lib64/stubs/libcuda.so
--     cudart library      : /public/apps/cuda/10.1/lib64/libcudart.so
--     cublas library      : /public/apps/cuda/10.1/lib64/libcublas.so
--     cufft library       : /public/apps/cuda/10.1/lib64/libcufft.so
--     curand library      : /public/apps/cuda/10.1/lib64/libcurand.so
--     cuDNN library       : /public/apps/cudnn/v7.6.5.32/cuda/lib64/libcudnn.so
--     nvrtc               : /public/apps/cuda/10.1/lib64/libnvrtc.so
--     CUDA include path   : /public/apps/cuda/10.1/include
--     NVCC executable     : /private/home/pbelevich/ccache/cuda/nvcc
--     NVCC flags          : -DONNX_NAMESPACE=onnx_torch;-gencode;arch=compute_60,code=sm_60;-Xcudafe;--diag_suppress=cc_clobber_ignored;-Xcudafe;--diag_suppress=integer_sign_change;-Xcudafe;--diag_suppress=useless_using_declaration;-Xcudafe;--diag_suppress=set_but_not_used;-Xcudafe;--diag_suppress=field_without_dll_interface;-Xcudafe;--diag_suppress=base_class_has_different_dll_interface;-Xcudafe;--diag_suppress=dll_interface_conflict_none_assumed;-Xcudafe;--diag_suppress=dll_interface_conflict_dllexport_assumed;-Xcudafe;--diag_suppress=implicit_return_from_non_void_function;-Xcudafe;--diag_suppress=unsigned_compare_with_zero;-Xcudafe;--diag_suppress=declared_but_not_referenced;-Xcudafe;--diag_suppress=bad_friend_decl;-std=c++14;-Xcompiler;-fPIC;--expt-relaxed-constexpr;--expt-extended-lambda;-Wno-deprecated-gpu-targets;--expt-extended-lambda;-gencode;arch=compute_60,code=sm_60;-Xcompiler;-fPIC;-DCUDA_HAS_FP16=1;-D__CUDA_NO_HALF_OPERATORS__;-D__CUDA_NO_HALF_CONVERSIONS__;-D__CUDA_NO_HALF2_OPERATORS__
--     CUDA host compiler  : /private/home/pbelevich/ccache/lib/cc
```

**Comment by malfet** [comment]
Only `CUDA-10.1 Update 1` has the necessary defines, which can be detected by checking `CUSPARSE_VER_MAJOR` >= 10

File: aten/src/ATen/cuda/CuSparseDescriptors.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+// guard the whole file
+#if !defined(_MSC_VER) && defined(__CUDACC__) && CUDART_VERSION >= 10010 // CUDA release >= 10.1 and not windows
```
Comment (malfet) [inline aten/src/ATen/cuda/CuSparseDescriptors.h:4]: @xwang233 please use something like && defined(CUSPARSE_VER_MAJOR) && CUSPARSE_VER_MAJOR >= 10 to detect CUDA-10.1 Update 1 or newer

Comment (xwang233) [reply aten/src/ATen/cuda/CuSparseDescriptors.h:4]: Thanks for the comments! I'll check that with different library versions.

**Comment by IvanYashchuk** [comment]
This PR got reverted in https://github.com/pytorch/pytorch/commit/6f8838cd2f6f189a1b89e9d0e13f029b5c97dc8f. @xwang233 what's the status of this work? Do you have plans on reviving it?

**Comment by xwang233** [comment]
Thanks @IvanYashchuk . Currently there is no update from my side. Feel free to restart a work on cuSPARSE descriptor class. Also feel free to reuse any code from this PR if you like.

**Pull request closed without merging**